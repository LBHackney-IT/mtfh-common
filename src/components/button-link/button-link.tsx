import React, { ComponentPropsWithoutRef, forwardRef } from 'react';
import cn from 'classnames';
import { navigateToUrl } from 'single-spa';

import './styles.scss';

export interface ButtonLinkProps extends ComponentPropsWithoutRef<'a'> {
  variant?: 'primary' | 'secondary';
  isDisabled?: boolean;
  isExternal?: boolean;
}

export const ButtonLink = forwardRef<HTMLAnchorElement, ButtonLinkProps>(
  function ButtonLink(
    {
      variant = 'primary',
      isExternal = false,
      href,
      isDisabled,
      children,
      className,
      ...props
    },
    ref
  ) {
    const buttonClasses = cn(
      'govuk-button',
      'lbh-button',
      {
        'govuk-secondary lbh-button--secondary': variant === 'secondary',
        'lbh-button--disabled govuk-button--disabled': isDisabled,
      },
      className
    );

    return (
      <a
        ref={ref}
        href={href}
        className={buttonClasses}
        onClick={!isExternal ? navigateToUrl : undefined}
        {...props}
      >
        {children}
      </a>
    );
  }
);
