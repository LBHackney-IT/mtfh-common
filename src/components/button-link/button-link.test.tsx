import React, { useEffect } from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { navigateToUrl } from 'single-spa';

import { ButtonLink } from './button-link';

jest.mock('single-spa');

test('it renders correctly', () => {
  const { container } = render(
    <ButtonLink href="http://localhost">Test Button</ButtonLink>
  );
  expect(container).toMatchSnapshot();
});

test('it renders the correct variant', () => {
  const { container } = render(
    <ButtonLink variant="secondary">Test Link</ButtonLink>
  );
  expect(container).toMatchSnapshot();
});

test('it can be made disabled', () => {
  const { container } = render(<ButtonLink isDisabled>Test Link</ButtonLink>);
  expect(container).toMatchSnapshot();
});

test('it uses internal navigation', () => {
  (navigateToUrl as jest.Mock).mockReset();
  render(<ButtonLink href="http://localhost">Test Button</ButtonLink>);
  const link = screen.getByText('Test Button');
  fireEvent.click(link);
  expect(navigateToUrl).toHaveBeenCalledTimes(1);
});

test('it does not use internal navigation when marked external', () => {
  (navigateToUrl as jest.Mock).mockReset();
  render(
    <ButtonLink href="http://localhost" isExternal>
      Test Button
    </ButtonLink>
  );
  const link = screen.getByText('Test Button');
  fireEvent.click(link);
  expect(navigateToUrl).toHaveBeenCalledTimes(0);
});

test('it accepts a ref', () => {
  const callback = jest.fn();
  const Comp = () => {
    const ref = React.createRef<HTMLAnchorElement>();

    useEffect(() => {
      if (ref.current) {
        callback(ref.current);
      }
    }, [ref]);
    return <ButtonLink ref={ref}>Test Button</ButtonLink>;
  };
  render(<Comp />);
  expect(callback).toBeCalledTimes(1);
});
