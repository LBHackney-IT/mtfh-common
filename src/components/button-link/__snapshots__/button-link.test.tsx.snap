// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it can be made disabled 1`] = `
<div>
  <a
    class="govuk-button lbh-button lbh-button--disabled govuk-button--disabled"
  >
    Test Link
  </a>
</div>
`;

exports[`it renders correctly 1`] = `
<div>
  <a
    class="govuk-button lbh-button"
    href="http://localhost"
  >
    Test Button
  </a>
</div>
`;

exports[`it renders the correct variant 1`] = `
<div>
  <a
    class="govuk-button lbh-button govuk-secondary lbh-button--secondary"
  >
    Test Link
  </a>
</div>
`;
